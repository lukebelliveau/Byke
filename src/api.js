// @flow
export const url = 'http://byke-graphql.herokuapp.com/graphql?query=';

const googleMapsAPIKey = 'AIzaSyAQhawlRtPkYdpC4VHgaB1n3yLOVOEZbYE';

const getAllStations = (stationParams: string) =>
  fetch(`${url}{allStations${stationParams}}`)
    .then(response => {
      return response.json().then(res => {
        return res.data.allStations;
      });
    })
    .catch(e => console.log(e));

const searchPlaces = (searchQuery: string, latitude, longitude) =>
  fetch(
    `https://maps.googleapis.com/maps/api/place/nearbysearch/json?key=${googleMapsAPIKey}&location=${latitude},${longitude}&radius=10000&keyword=${searchQuery}`
  )
    .then(response => {
      response.json().then(json => json.results);
    })
    .catch(e => console.log(e));

export default {
  getAllStations,
  searchPlaces,
};
