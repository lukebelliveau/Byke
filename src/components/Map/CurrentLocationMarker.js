import React from 'react';
import { View, Keyboard } from 'react-native';
import MapView from 'react-native-maps';

const CurrentLocationMarker = ({ coordinate }) =>
  <MapView.Marker
    coordinate={coordinate}
    title="Your location"
    onPress={Keyboard.dismiss}
  >
    <View style={style} />
  </MapView.Marker>;

const style = {
  width: 20,
  height: 20,
  borderRadius: 50,
  backgroundColor: 'dodgerblue',
  borderColor: 'white',
  borderWidth: 1,
};

export default CurrentLocationMarker;
