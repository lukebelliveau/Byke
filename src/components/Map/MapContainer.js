// @flow
import { connect } from 'react-redux';
import MapComponent from './Map';
import actions from '../../redux/actions';

const mapDispatchToProps = dispatch => ({
  loadingStarted: () => {
    dispatch(actions.loadingStarted());
  },
  loadingFinished: () => {
    dispatch(actions.loadingFinished());
  },
  stationsFetched: stations => {
    dispatch(actions.stationsFetched(stations));
  },
  locationUpdated: location => {
    dispatch(actions.locationUpdated(location));
  },
});

const mapStateToProps = state => ({
  stations: state.stations,
  trip: state.trip,
  currentLocation: state.currentLocation,
  region: state.region,
});

const Map = connect(mapStateToProps, mapDispatchToProps)(MapComponent);

export default Map;
