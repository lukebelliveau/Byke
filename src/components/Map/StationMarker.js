// @flow
import React from 'react';
import MapView from 'react-native-maps';
import { Keyboard } from 'react-native';
import utils from '../../utils';

import Circle from '../shared/Circle';

const Marker = ({
  coordinate,
  stationName,
  currentLocation,
  availableBikes,
  availableDocks,
}: {
  coordinate: Object,
  stationName: string,
  availableBikes: number,
}) => {
  return (
    <MapView.Marker
      coordinate={coordinate}
      title={stationName}
      description={`${availableBikes} bikes available, ${availableDocks} docks available`}
      onPress={Keyboard.dismiss}
      onCalloutPress={() =>
        utils.displayNavigationAlert(
          currentLocation,
          stationName,
          coordinate,
          availableBikes,
          availableDocks
        )}
    >
      <Circle color={availableBikes > 5 ? 'green' : 'red'} />
    </MapView.Marker>
  );
};

export default Marker;
