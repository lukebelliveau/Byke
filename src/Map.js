// @flow
import React, { Component } from 'react';
import { Keyboard, View, AsyncStorage } from 'react-native';
import MapView from 'react-native-maps';

import StationMarker from './StationMarker';
import api from './api';
import Loading from './Loading';

type Station = {
  latitude: number,
  longitude: number,
  stationName: string,
  availableBikes: number,
};

type Props = {
  region: { number: string },
  stations: Array<Station>,
  style: Object,
  trip: Object,
};

const initialState = {
  stations: [],
};
class Map extends Component {
  state = initialState;
  constructor(props: Props) {
    super(props);

    this.props = props;
  }

  componentDidMount() {
    this.fetchStationInfo();
  }

  componentDidUpdate() {
    this.map.animateToRegion(this.props.region)
  }

  loadStations() {
    const stations = AsyncStorage.getItem('@Byke:stations')
      .then(stations => {
        if (stations !== null) {
          this.setState({
            stations: JSON.parse(stations),
          });
        } else {
          this.fetchStationInfo();
        }
      })
      .catch(e => {
        Error('error fetching stations!' + e);
      });
  }

  fetchStationInfo = () => {
    api
      .getAllStations(`{ stationName, availableBikes, latitude, longitude }`)
      .then(stations => {
        AsyncStorage.setItem(
          '@Byke:stations',
          JSON.stringify(stations)
        ).catch(e => console.log('error:' + e));
        this.setState({
          stations,
        });
      });
  };

  render() {
    const trip = this.props.trip;
    const region = this.props.region;
    const stations = this.state.stations;

    return (
      <View style={{ flex: 1 }}>
        {
          <MapView
            style={{ flex: 1 }}
            //region={region}
            initialRegion={region}
            onPress={Keyboard.dismiss}
            ref={ref => { this.map = ref; }}
          >
            {trip
              ? <View>
                  <MapView.Marker
                    coordinate={trip.currentLocation}
                    pinColor="blue"
                  />
                  <MapView.Marker coordinate={trip.destination} />
                </View>
              : <MapView.Marker coordinate={region} />}

            {stations.map((station, index) =>
              <StationMarker
                testId={station.stationName}
                stationName={station.stationName}
                coordinate={station}
                availableBikes={station.availableBikes}
                key={index}
              />
            )}
          </MapView>
        }
        {stations.length === 0 ? <Loading /> : null}
      </View>
    );
  }
}
export default Map;
