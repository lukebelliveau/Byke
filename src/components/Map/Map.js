// @flow
import React, { Component } from 'react';
import { Keyboard, View } from 'react-native';
import MapView from 'react-native-maps';

import StationMarker from './StationMarker';
import api from '../../api';
import { Region, Location, Station, Trip } from '../../Types';

type Props = {
  stations: Array<Station>,
  region: Region,
  trip: Trip,
  loadingStarted: () => void,
  loadingFinished: () => void,
  locationUpdated: Location => void,
  stationsFetched: (Array<Station>) => void,
};

const positionError = error => {
  throw Error(error.message);
};

const positionOptions = {
  timeout: Number.POSITIVE_INFINITY,
  maximumAge: 0,
  enableHighAccuracy: false,
};

class Map extends Component {
  map: {
    animateToRegion: Region => void,
  };
  constructor(props: Props) {
    super(props);

    this.props = props;
    this.fetchStationInfo();
    this.watchLocation();
  }

  watchLocation() {
    navigator.geolocation.watchPosition(
      location => {
        this.props.locationUpdated(location.coords), null, {};
      },
      positionError,
      positionOptions
    );
  }

  componentDidUpdate() {
    this.map.animateToRegion(this.props.region);
  }

  fetchStationInfo = () => {
    this.props.loadingStarted();
    api
      .getAllStations(
        `{ stationName, availableBikes, availableDocks, latitude, longitude }`
      )
      .then(stations => {
        this.props.stationsFetched(stations);
        this.props.loadingFinished();
      });
  };

  render() {
    const trip = this.props.trip;
    const region = this.props.region;
    const stations = this.props.stations;
    const currentLocation = this.props.currentLocation;

    return (
      <View style={{ flex: 1 }}>
        {
          <MapView
            style={{ flex: 1 }}
            onPress={Keyboard.dismiss}
            ref={ref => {
              this.map = ref;
            }}
          >
            {trip
              ? <View>
                  <MapView.Marker
                    coordinate={currentLocation}
                    pinColor="blue"
                  />
                  <MapView.Marker coordinate={trip.destination} />
                </View>
              : <MapView.Marker coordinate={currentLocation} />}

            {stations.map((station, index) => {
              return (
                <StationMarker
                  testId={station.stationName}
                  stationName={station.stationName}
                  coordinate={station}
                  currentLocation={currentLocation}
                  availableBikes={station.availableBikes}
                  availableDocks={station.availableDocks}
                  key={index}
                />
              );
            })}
          </MapView>
        }
      </View>
    );
  }
}

export default Map;
