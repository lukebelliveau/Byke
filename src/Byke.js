// @flow
import React from 'react';
import { StyleSheet, View } from 'react-native';

import Map from './Map/MapContainer';
import SearchBar from './SearchBar/SearchBarContainer';
import LocationList from './SearchBar/LocationListContainer';
import Loading from './Loading';

class Byke extends React.Component {
  constructor() {
    super();
  }

  render() {
    return (
      <View style={styles.container}>
        <SearchBar style={styles.destination} />
        <View style={styles.map}>
          <Map />
          <LocationList />
          <Loading />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  destination: {
    flex: 1,
  },
  map: {
    flex: 6,
  },
});

export default Byke;
