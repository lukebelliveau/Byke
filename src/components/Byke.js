// @flow
import React from 'react';
import { StyleSheet, View } from 'react-native';

import Map from './Map/MapContainer';
import SearchBar from './SearchBar/SearchBarContainer';
import LocationList from './SearchBar/LocationListContainer';
import Loading from './shared/Loading';

class Byke extends React.Component {
  constructor() {
    super();
  }

  render() {
    return (
      <View style={styles.container}>
        <SearchBar />
        <View style={styles.map}>
          <Map />
          <LocationList />
          <Loading />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  map: {
    flex: 6,
  },
});

export default Byke;
