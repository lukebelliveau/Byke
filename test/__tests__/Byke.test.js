import React from 'react';
import { shallow } from 'enzyme';
import { createWaitForElement } from 'enzyme-wait';

import api from '../../src/api';
import Byke from '../../src/Byke';
import Map from '../../src/Map';

api.searchPlaces = jest.fn(() => new Promise(resolve => resolve([])));

it('calls searchPlaces API with location when user submits destination', done => {
  const latitude = 50;
  const longitude = 50;
  mockPosition(latitude, longitude);
  api.getAllStations = jest.fn(() => new Promise(resolve => resolve([])));
  const waitForMap = createWaitForElement(Map);
  const searchQuery = 'DisneyWorld';
  const byke = shallow(<Byke />);
  byke.instance().componentDidMount();

  waitForMap(byke).then(component => {
    component.instance().searchDestination(searchQuery);

    expect(api.searchPlaces).toBeCalledWith(searchQuery, latitude, longitude);
    done();
  });
});

/*
 * CONFIG
 * */
const mockPosition = (latitude, longitude) => {
  const mockGetCurrentPosition = callback => {
    callback({
      coords: {
        latitude,
        longitude,
      },
    });
  };

  const mockGeolocation = {
    getCurrentPosition: jest.fn(mockGetCurrentPosition),
    watchPosition: jest.fn(),
  };

  global.navigator.geolocation = mockGeolocation;
};

const mockGeolocation = {
  getCurrentPosition: jest.fn(),
  watchPosition: jest.fn(),
};

global.navigator.geolocation = mockGeolocation;
