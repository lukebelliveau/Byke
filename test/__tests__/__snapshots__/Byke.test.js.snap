// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runs 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "dodgerblue",
          "elevation": 10,
          "flexDirection": "column-reverse",
          "shadowColor": "black",
          "shadowOpacity": 100,
          "shadowRadius": 10,
        },
      ]
    }
  >
    <TextInput
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Where are you going?"
      placeholderTextColor="lightgray"
      returnKeyType="go"
      style={
        Object {
          "backgroundColor": "deepskyblue",
          "borderRadius": 10,
          "color": "white",
          "flex": 1,
          "fontSize": 15,
          "margin": 20,
          "maxHeight": 25,
          "padding": 5,
          "paddingLeft": 10,
          "width": "75%",
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "flex": 6,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <MapView
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View>
          <Marker
            coordinate={
              Object {
                "latitude": 0,
                "latitudeDelta": 0.01,
                "longitude": 0,
                "longitudeDelta": 0.01,
              }
            }
          />
        </View>
      </MapView>
    </View>
    <View
      pointerEvents="none"
      style={
        Object {
          "display": "flex",
          "height": "100%",
          "marginRight": "10%",
          "paddingLeft": "5%",
          "paddingRight": "5%",
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <RCTScrollView
        showsVerticalScrollIndicator={false}
      >
        <View />
      </RCTScrollView>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "center",
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 20,
            "bottom": 20,
            "display": "flex",
            "elevation": 5,
            "justifyContent": "center",
            "padding": 10,
            "position": "absolute",
            "width": "50%",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          Loading...
        </Text>
      </View>
    </View>
  </View>
</View>
`;
